// ngex/bundles/ngex.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('rxjs'), require('@angular/platform-browser')) :
    typeof define === 'function' && define.amd ? define('ngex', ['exports', '@angular/core', 'rxjs', '@angular/platform-browser'], factory) :
    (factory((global.ngex = {}),global.ng.core,global.rxjs,global.ng.platformBrowser));
}(this, (function (exports,i0,rxjs,platformBrowser) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    /** @type {?} */
    var NgexServiceConfig = new i0.InjectionToken('NgexStore');

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var Ngex = /** @class */ (function () {
        function Ngex(config) {
            var _this = this;
            this.config = config;
            this.storex = {};
            this.actions = [];
            this.statepv = new rxjs.BehaviorSubject(this.storex);
            this.store = this.statepv.asObservable();
            this.state = this.store.subscribe(function (state) { _this.state = state; });
            if (config) {
                this.createStore(config.store);
                this.createActions(config.actions);
            }
        }
        /**
         * @param {?} store
         * @return {?}
         */
        Ngex.prototype.createStore = /**
         * @param {?} store
         * @return {?}
         */
            function (store) {
                this.storex = store;
                this.statepv.next(this.storex);
            };
        /**
         * @param {?} actions
         * @return {?}
         */
        Ngex.prototype.createActions = /**
         * @param {?} actions
         * @return {?}
         */
            function (actions) {
                this.actions = actions(this.storex);
                this.statepv.next(this.storex);
            };
        /**
         * @param {?} type
         * @param {?=} val
         * @return {?}
         */
        Ngex.prototype.emit = /**
         * @param {?} type
         * @param {?=} val
         * @return {?}
         */
            function (type, val) {
                if (val === void 0) {
                    val = null;
                }
                this.actions.find(function (a) {
                    if (a.type === type) {
                        return a.action(val);
                    }
                });
                this.statepv.next(this.storex);
            };
        Ngex.decorators = [
            { type: i0.Injectable, args: [{
                        providedIn: 'root'
                    },] }
        ];
        /** @nocollapse */
        Ngex.ctorParameters = function () {
            return [
                { type: undefined, decorators: [{ type: i0.Inject, args: [NgexServiceConfig,] }] }
            ];
        };
        /** @nocollapse */ Ngex.ngInjectableDef = i0.defineInjectable({ factory: function Ngex_Factory() { return new Ngex(i0.inject(NgexServiceConfig)); }, token: Ngex, providedIn: "root" });
        return Ngex;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var NgexModule = /** @class */ (function () {
        function NgexModule(parentModule) {
            if (parentModule) {
                throw new Error('NgexModule is already loaded. Import it in the AppModule only');
            }
        }
        /**
         * @param {?} config
         * @return {?}
         */
        NgexModule.forRoot = /**
         * @param {?} config
         * @return {?}
         */
            function (config) {
                return {
                    ngModule: NgexModule,
                    providers: [
                        Ngex, {
                            provide: NgexServiceConfig,
                            useValue: config
                        }
                    ]
                };
            };
        NgexModule.decorators = [
            { type: i0.NgModule, args: [{ imports: [platformBrowser.BrowserModule], declarations: [], exports: [] },] }
        ];
        /** @nocollapse */
        NgexModule.ctorParameters = function () {
            return [
                { type: NgexModule, decorators: [{ type: i0.Optional }, { type: i0.SkipSelf }] }
            ];
        };
        return NgexModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    exports.Ngex = Ngex;
    exports.NgexModule = NgexModule;
    exports.ɵa = NgexServiceConfig;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
