// ngex/fesm2015/ngex.js
import { InjectionToken, Injectable, Inject, NgModule, Optional, SkipSelf, defineInjectable, inject } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { BrowserModule } from '@angular/platform-browser';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/** @type {?} */
const NgexServiceConfig = new InjectionToken('NgexStore');

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
class Ngex {
    /**
     * @param {?} config
     */
    constructor(config) {
        this.config = config;
        this.storex = {};
        this.actions = [];
        this.statepv = new BehaviorSubject(this.storex);
        this.store = this.statepv.asObservable();
        this.state = this.store.subscribe((state) => { this.state = state; });
        if (config) {
            this.createStore(config.store);
            this.createActions(config.actions);
        }
    }
    /**
     * @param {?} store
     * @return {?}
     */
    createStore(store) {
        this.storex = store;
        this.statepv.next(this.storex);
    }
    /**
     * @param {?} actions
     * @return {?}
     */
    createActions(actions) {
        this.actions = actions(this.storex);
        this.statepv.next(this.storex);
    }
    /**
     * @param {?} type
     * @param {?=} val
     * @return {?}
     */
    emit(type, val = null) {
        this.actions.find((a) => {
            if (a.type === type) {
                return a.action(val);
            }
        });
        this.statepv.next(this.storex);
    }
}
Ngex.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
Ngex.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [NgexServiceConfig,] }] }
];
/** @nocollapse */ Ngex.ngInjectableDef = defineInjectable({ factory: function Ngex_Factory() { return new Ngex(inject(NgexServiceConfig)); }, token: Ngex, providedIn: "root" });

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
class NgexModule {
    /**
     * @param {?} parentModule
     */
    constructor(parentModule) {
        if (parentModule) {
            throw new Error('NgexModule is already loaded. Import it in the AppModule only');
        }
    }
    /**
     * @param {?} config
     * @return {?}
     */
    static forRoot(config) {
        return {
            ngModule: NgexModule,
            providers: [
                Ngex, {
                    provide: NgexServiceConfig,
                    useValue: config
                }
            ]
        };
    }
}
NgexModule.decorators = [
    { type: NgModule, args: [{ imports: [BrowserModule], declarations: [], exports: [] },] }
];
/** @nocollapse */
NgexModule.ctorParameters = () => [
    { type: NgexModule, decorators: [{ type: Optional }, { type: SkipSelf }] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

export { Ngex, NgexModule, NgexServiceConfig as ɵa };
