// ngex/esm2015/lib/ngex.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { NgModule, Optional, SkipSelf } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import { Ngex } from './ngex.service';
import { NgexServiceConfig } from './ngex.config';
export class NgexModule {
    /**
     * @param {?} parentModule
     */
    constructor(parentModule) {
        if (parentModule) {
            throw new Error('NgexModule is already loaded. Import it in the AppModule only');
        }
    }
    /**
     * @param {?} config
     * @return {?}
     */
    static forRoot(config) {
        return {
            ngModule: NgexModule,
            providers: [
                Ngex, {
                    provide: NgexServiceConfig,
                    useValue: config
                }
            ]
        };
    }
}
NgexModule.decorators = [
    { type: NgModule, args: [{ imports: [BrowserModule], declarations: [], exports: [] },] }
];
/** @nocollapse */
NgexModule.ctorParameters = () => [
    { type: NgexModule, decorators: [{ type: Optional }, { type: SkipSelf }] }
];
