// ngex/esm2015/lib/ngex.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable, Inject } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { NgexServiceConfig } from './ngex.config';
import * as i0 from "@angular/core";
import * as i1 from "./ngex.config";
export class Ngex {
    /**
     * @param {?} config
     */
    constructor(config) {
        this.config = config;
        this.storex = {};
        this.actions = [];
        this.statepv = new BehaviorSubject(this.storex);
        this.store = this.statepv.asObservable();
        this.state = this.store.subscribe((state) => { this.state = state; });
        if (config) {
            this.createStore(config.store);
            this.createActions(config.actions);
        }
    }
    /**
     * @param {?} store
     * @return {?}
     */
    createStore(store) {
        this.storex = store;
        this.statepv.next(this.storex);
    }
    /**
     * @param {?} actions
     * @return {?}
     */
    createActions(actions) {
        this.actions = actions(this.storex);
        this.statepv.next(this.storex);
    }
    /**
     * @param {?} type
     * @param {?=} val
     * @return {?}
     */
    emit(type, val = null) {
        this.actions.find((a) => {
            if (a.type === type) {
                return a.action(val);
            }
        });
        this.statepv.next(this.storex);
    }
}
Ngex.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
Ngex.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [NgexServiceConfig,] }] }
];
/** @nocollapse */ Ngex.ngInjectableDef = i0.defineInjectable({ factory: function Ngex_Factory() { return new Ngex(i0.inject(i1.NgexServiceConfig)); }, token: Ngex, providedIn: "root" });
if (false) {
    /** @type {?} */
    Ngex.prototype.storex;
    /** @type {?} */
    Ngex.prototype.actions;
    /** @type {?} */
    Ngex.prototype.statepv;
    /** @type {?} */
    Ngex.prototype.store;
    /** @type {?} */
    Ngex.prototype.state;
    /** @type {?} */
    Ngex.prototype.config;
}
