// ngex/esm5/lib/ngex.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable, Inject } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { NgexServiceConfig } from './ngex.config';
import * as i0 from "@angular/core";
import * as i1 from "./ngex.config";
var Ngex = /** @class */ (function () {
    function Ngex(config) {
        var _this = this;
        this.config = config;
        this.storex = {};
        this.actions = [];
        this.statepv = new BehaviorSubject(this.storex);
        this.store = this.statepv.asObservable();
        this.state = this.store.subscribe(function (state) { _this.state = state; });
        if (config) {
            this.createStore(config.store);
            this.createActions(config.actions);
        }
    }
    /**
     * @param {?} store
     * @return {?}
     */
    Ngex.prototype.createStore = /**
     * @param {?} store
     * @return {?}
     */
    function (store) {
        this.storex = store;
        this.statepv.next(this.storex);
    };
    /**
     * @param {?} actions
     * @return {?}
     */
    Ngex.prototype.createActions = /**
     * @param {?} actions
     * @return {?}
     */
    function (actions) {
        this.actions = actions(this.storex);
        this.statepv.next(this.storex);
    };
    /**
     * @param {?} type
     * @param {?=} val
     * @return {?}
     */
    Ngex.prototype.emit = /**
     * @param {?} type
     * @param {?=} val
     * @return {?}
     */
    function (type, val) {
        if (val === void 0) { val = null; }
        this.actions.find(function (a) {
            if (a.type === type) {
                return a.action(val);
            }
        });
        this.statepv.next(this.storex);
    };
    Ngex.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    Ngex.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [NgexServiceConfig,] }] }
    ]; };
    /** @nocollapse */ Ngex.ngInjectableDef = i0.defineInjectable({ factory: function Ngex_Factory() { return new Ngex(i0.inject(i1.NgexServiceConfig)); }, token: Ngex, providedIn: "root" });
    return Ngex;
}());
export { Ngex };
if (false) {
    /** @type {?} */
    Ngex.prototype.storex;
    /** @type {?} */
    Ngex.prototype.actions;
    /** @type {?} */
    Ngex.prototype.statepv;
    /** @type {?} */
    Ngex.prototype.store;
    /** @type {?} */
    Ngex.prototype.state;
    /** @type {?} */
    Ngex.prototype.config;
}
