// ngex/fesm5/ngex.js
import { InjectionToken, Injectable, Inject, NgModule, Optional, SkipSelf, defineInjectable, inject } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { BrowserModule } from '@angular/platform-browser';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/** @type {?} */
var NgexServiceConfig = new InjectionToken('NgexStore');

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var Ngex = /** @class */ (function () {
    function Ngex(config) {
        var _this = this;
        this.config = config;
        this.storex = {};
        this.actions = [];
        this.statepv = new BehaviorSubject(this.storex);
        this.store = this.statepv.asObservable();
        this.state = this.store.subscribe(function (state) { _this.state = state; });
        if (config) {
            this.createStore(config.store);
            this.createActions(config.actions);
        }
    }
    /**
     * @param {?} store
     * @return {?}
     */
    Ngex.prototype.createStore = /**
     * @param {?} store
     * @return {?}
     */
    function (store) {
        this.storex = store;
        this.statepv.next(this.storex);
    };
    /**
     * @param {?} actions
     * @return {?}
     */
    Ngex.prototype.createActions = /**
     * @param {?} actions
     * @return {?}
     */
    function (actions) {
        this.actions = actions(this.storex);
        this.statepv.next(this.storex);
    };
    /**
     * @param {?} type
     * @param {?=} val
     * @return {?}
     */
    Ngex.prototype.emit = /**
     * @param {?} type
     * @param {?=} val
     * @return {?}
     */
    function (type, val) {
        if (val === void 0) { val = null; }
        this.actions.find(function (a) {
            if (a.type === type) {
                return a.action(val);
            }
        });
        this.statepv.next(this.storex);
    };
    Ngex.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    Ngex.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [NgexServiceConfig,] }] }
    ]; };
    /** @nocollapse */ Ngex.ngInjectableDef = defineInjectable({ factory: function Ngex_Factory() { return new Ngex(inject(NgexServiceConfig)); }, token: Ngex, providedIn: "root" });
    return Ngex;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var NgexModule = /** @class */ (function () {
    function NgexModule(parentModule) {
        if (parentModule) {
            throw new Error('NgexModule is already loaded. Import it in the AppModule only');
        }
    }
    /**
     * @param {?} config
     * @return {?}
     */
    NgexModule.forRoot = /**
     * @param {?} config
     * @return {?}
     */
    function (config) {
        return {
            ngModule: NgexModule,
            providers: [
                Ngex, {
                    provide: NgexServiceConfig,
                    useValue: config
                }
            ]
        };
    };
    NgexModule.decorators = [
        { type: NgModule, args: [{ imports: [BrowserModule], declarations: [], exports: [] },] }
    ];
    /** @nocollapse */
    NgexModule.ctorParameters = function () { return [
        { type: NgexModule, decorators: [{ type: Optional }, { type: SkipSelf }] }
    ]; };
    return NgexModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

export { Ngex, NgexModule, NgexServiceConfig as ɵa };
