// ngex/__ivy_ngcc__/fesm2015/ngex.js
import { InjectionToken, Injectable, Inject, NgModule, Optional, SkipSelf, defineInjectable, inject } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { BrowserModule } from '@angular/platform-browser';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/** @type {?} */
import * as ɵngcc0 from '@angular/core';
const NgexServiceConfig = new InjectionToken('NgexStore');

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
class Ngex {
    /**
     * @param {?} config
     */
    constructor(config) {
        this.config = config;
        this.storex = {};
        this.actions = [];
        this.statepv = new BehaviorSubject(this.storex);
        this.store = this.statepv.asObservable();
        this.state = this.store.subscribe((state) => { this.state = state; });
        if (config) {
            this.createStore(config.store);
            this.createActions(config.actions);
        }
    }
    /**
     * @param {?} store
     * @return {?}
     */
    createStore(store) {
        this.storex = store;
        this.statepv.next(this.storex);
    }
    /**
     * @param {?} actions
     * @return {?}
     */
    createActions(actions) {
        this.actions = actions(this.storex);
        this.statepv.next(this.storex);
    }
    /**
     * @param {?} type
     * @param {?=} val
     * @return {?}
     */
    emit(type, val = null) {
        this.actions.find((a) => {
            if (a.type === type) {
                return a.action(val);
            }
        });
        this.statepv.next(this.storex);
    }
}
Ngex.ɵfac = function Ngex_Factory(t) { return new (t || Ngex)(ɵngcc0.ɵɵinject(NgexServiceConfig)); };
Ngex.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: Ngex, factory: Ngex.ɵfac, providedIn: 'root' });
/** @nocollapse */
Ngex.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [NgexServiceConfig,] }] }
];
/** @nocollapse */ Ngex.ngInjectableDef = defineInjectable({ factory: function Ngex_Factory() { return new Ngex(inject(NgexServiceConfig)); }, token: Ngex, providedIn: "root" });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(Ngex, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: undefined, decorators: [{
                type: Inject,
                args: [NgexServiceConfig]
            }] }]; }, null); })();

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
class NgexModule {
    /**
     * @param {?} parentModule
     */
    constructor(parentModule) {
        if (parentModule) {
            throw new Error('NgexModule is already loaded. Import it in the AppModule only');
        }
    }
    /**
     * @param {?} config
     * @return {?}
     */
    static forRoot(config) {
        return {
            ngModule: NgexModule,
            providers: [
                Ngex, {
                    provide: NgexServiceConfig,
                    useValue: config
                }
            ]
        };
    }
}
NgexModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgexModule });
NgexModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgexModule_Factory(t) { return new (t || NgexModule)(ɵngcc0.ɵɵinject(NgexModule, 12)); }, imports: [[BrowserModule]] });
/** @nocollapse */
NgexModule.ctorParameters = () => [
    { type: NgexModule, decorators: [{ type: Optional }, { type: SkipSelf }] }
];
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgexModule, { imports: function () { return [BrowserModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgexModule, [{
        type: NgModule,
        args: [{ imports: [BrowserModule], declarations: [], exports: [] }]
    }], function () { return [{ type: NgexModule, decorators: [{
                type: Optional
            }, {
                type: SkipSelf
            }] }]; }, null); })();

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

export { Ngex, NgexModule, NgexServiceConfig as ɵa };
